pex/pex.py

```diff
@@ -640,26 +657,29 @@ def execute_interpreter(self):
             code.interact()
             return None
 
-    def execute_with_options(self, options, args):
+    def execute_with_options(
+        self,
+        python_options,  # type: List[str]
+        args,  # List[str]
+    ):
+        # type: (...) -> Union[NoReturn, Any]
         """
         Restart the process passing the given options to the python interpreter
         """
-        # Find the pex bootstrap location
-        # We expect this environment variable to be set by __main__.py
+        # Find the installed (unzipped) PEX entry point.
         main = sys.modules.get("__main__")
-        if main:
-            run_pex = os.path.dirname(main.__file__)
-            interpreter = PythonInterpreter.get().binary
-            cmdline = [interpreter] + options + [run_pex] + args
-            TRACER.log(
-                "Re-executing with interpreter options: "
-                "cmdline={cmdline!r}, ".format(
-                    cmdline=" ".join(cmdline),
-                )
+        if not main or not main.__file__:
+            # N.B.: This should never happen.
+            return "Unable to resolve PEX __main__ module file: {}".format(main)
```

Comment:
How about raising `AssertionError`?

Reply:
As the tests show, those can be turned off. This gives a clean non-zero exit with a non-backtrace mess message and is in line with all other exits here. What are you seeing as the advantage of an assertion over this? Collapsing a comment?

Reply:
That usually when you say "this shouldn't happen", using an assert is the way to ensure that doesn't happen. Sg though

Reply:
Well, an assert doesn't "ensure it doesn't happen" as the -O tests show, I think you must mean it expresses intent better, or at least more concisely. That I don't disagree with, but will leave as Loren established it.